packages/app/src/server/routes/apiv3/page-listing.ts

```diff
@@ -118,7 +118,8 @@ const routerFactory = (crowi: Crowi): Router => {
     const pageService: PageService = crowi.pageService!;
 
     try {
-      const pages = pageIds != null
+      const pageIds = Object.values(pageIdsObj ?? {});
+      const pages = pageIds != null && pageIds.length > 0
```

Comment:
page-listing/info のエラーは、req.query.pageIds が `{ "1": "{ObjectId}" }` という形式になっており、引数として配列を取る Page.findByIdsAndViewer にこのオブジェクトを入れていたことが原因でした。

クライアントからは配列で渡していることは確認していますが、req.query.pageIds がオブジェクトになってしまうという根本原因の解決にはなっていません。このコードで良いか別の方法を探るべきかの判断をお願いしたいです。

Reply:
根本解決が必要。問題の切り分けを細かくやってください。

Reply:
## 原因
express 内部でクエリパラメーターを解析するために qs というモジュールを使っている。クライアントから渡されたクエリパラメーターは qs.parse というメソッドを使って解析している。qs では配列は20に制限されている。20以上の配列が渡されるとインデックスをキーとするオブジェクトに変換されるためエラーが出ていた。

- [issue](https://github.com/expressjs/body-parser/issues/289)
- [express がクエリパラメーターを解析している箇所](https://github.com/expressjs/express/blob/508936853a6e311099c9985d4c11a4b1b8f6af07/lib/middleware/query.js#L15-L47)
- [harb/qs](https://github.com/ljharb/qs)
   -  > qs will also limit specifying indices in an array to a maximum index of 20. Any array members with an index of greater than 20 will instead be converted to an object with the index as the key. This is needed to handle cases when someone sent, for example, a[999999999] and it will take significant time to iterate over this huge array.

## 解決方法
pageIds は JSON.stringify して文字列で渡すようにしました。

Reply:
express への middleware の整理で対応できないかな？
20件以上になるかどうかを API ごとに考慮して JSON.stringify するようなガイドラインを設けたくない。

Reply:
https://github.com/weseek/growi/pull/7373/commits/783aef84877e53793f9e201bc3d606c80ea5bf63 で対応しました